### PR TITLE
#10 replace unix socket by tcp

### PR DESCRIPTION
**Pull request by ale64bit**
Trying to build on the comment:

> I want to move off of the unix socket and want to listen on a TCP socket instead. I am planning to listen on a random port and keep the port process is listening under $HOME/.gops/.addr. This will also enable the windows support.

Not sure about the perms, though.

### CODE REVIEW

**Comment by ale64bit** [comment]
I signed it!

File: agent/agent.go [hunk 1]
```diff
@@ -64,6 +76,17 @@ func Start() error {
 	return err
 }
 
+func configdir() (string, error) {
+	gopsdir := fmt.Sprintf("%s/.gops", os.Getenv("HOME"))
```
Comment (rakyll) [inline agent/agent.go]: Let's put things under .config/gops on Unix-like systems and under APPDATA on windows.

```
func configdir() (string, error) {
	if runtime.GOOS == "windows" {
		return filepath.Join(os.Getenv("APPDATA"), "gops"), nil
	}
	homeDir := guessUnixHomeDir()
	if homeDir == "" {
		return "", errors.New("unable to get current user home directory: os/user lookup failed; $HOME is empty")
	}
	return filepath.Join(homeDir, ".config", "gops"), nil
}

func guessUnixHomeDir() string {
	usr, err := user.Current()
	if err == nil {
		return usr.HomeDir
	}
	return os.Getenv("HOME")
}
```

File: agent/agent.go [hunk 2]
```diff
@@ -64,6 +76,17 @@ func Start() error {
...
+	if _, err := os.Stat(gopsdir); os.IsNotExist(err) {
+		err = os.Mkdir(gopsdir, os.ModePerm)
```
Comment (rakyll) [inline agent/agent.go]: Use os.MkdirAll. It is similar to mkdir -p.

File: cmd.go [hunk 3]
```diff
@@ -106,9 +107,22 @@ func vitals(pid int) error {
 	return nil
 }
 
+func getport(pid int) (string, error) {
+	portfile := fmt.Sprintf("%s/.gops/.%d", os.Getenv("HOME"), pid)
```
Comment (rakyll) [inline cmd.go]: use configdir to find out the config directory.

File: agent/agent.go [hunk 4]
```diff
@@ -21,14 +23,24 @@ import (
...
 	// TODO(jbd): Expose these endpoints on HTTP. Then, we can enable
 	// the agent on Windows systems.
-	sock := fmt.Sprintf("/tmp/gops%d.sock", os.Getpid())
-	l, err := net.Listen("unix", sock)
+	l, err := net.Listen("tcp", "127.0.0.1:0")
+	if err != nil {
+		return err
+	}
+	port := l.Addr().(*net.TCPAddr).Port
+	portfile := fmt.Sprintf("%s/.%d", gopsdir, os.Getpid())
```
Comment (rakyll) [inline agent/agent.go]: Filename doesn't have to be hidden.

portfile := fmt.Sprintf("%s/%d", gopsdir, os.Getpid())

is fine

**Comment by ale64bit** [comment]
thx for comments. PR updated.

**Comment by rakyll** [comment]
Thanks!

**Comment by ale64bit** [comment]
@rakyll No problem :) I would like, if you are not against, to contribute further. May I ask, what scope and roadmap do you have in mind for this tool?

**Pull request merged**